query_optimizer/OptimizerTree.hpp

```diff
@@ -118,6 +122,169 @@ class OptimizerTree : public OptimizerTreeBase {
+        : key_(key), value_(value) {  }
+
+    std::string getName() const {
+      return "ProtoProperty";
+    }
+
+    void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                             std::vector<std::string> *inline_field_values,
+                             std::vector<std::string> *non_container_child_field_names,
+                             std::vector<TreeNodeType> *non_container_child_fields,
+                             std::vector<std::string> *container_child_field_names,
+                             std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+      inline_field_names->push_back(std::string("Property"));
+      inline_field_values->push_back(key_);
+
+      inline_field_names->push_back(std::string("Value"));
+      inline_field_values->push_back(value_);
+    }
+
+   private:
+    std::string key_;
+    std::string value_;
+  };  // class OptimizerProtoRepresentation
+
+ public:
+  OptimizerProtoRepresentation() : properties_() { }
+
+  std::string getName() const {
+    return "ProtoDescription";
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value A string value.
+   */
+  void addProperty(std::string key, std::string value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, value));
+    properties_.push_back(property);
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value An int value which is converted into a string.
+   */
+  void addProperty(std::string key, int value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, std::to_string(value)));
+    properties_.push_back(property);
+  }
+
+  void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                           std::vector<std::string> *inline_field_values,
+                           std::vector<std::string> *non_container_child_field_names,
+                           std::vector<TreeNodeType> *non_container_child_fields,
+                           std::vector<std::string> *container_child_field_names,
+                           std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+    for (auto property : properties_) {
+      non_container_child_field_names->push_back("Property");
+      non_container_child_fields->push_back(property);
+    }
+  }
+  // A list of managed properties.
+  std::vector<std::shared_ptr<const OptimizerProtoPropertyRepresentation> > properties_;
```

Comment:
@saketj I think one difference is that PtrList (as well as PtrVector) can do ownership transitions. You will see that many classes in Quickstep have this macro at the bottom:
`DISALLOW_COPY_AND_ASSIGN(ClassName)`

This will disallow copy assignment (to avoid potential expensive copy assignment that we are not aware of), so you cannot push an existing object into a vector, say
```
A a; // Suppose A has the DISALLOW_COPY_AND_ASSIGN(A) macro
vector<A> a_vec;
a_vec.push_back(a); // Compile error
a_vec.push_back(std::move(a)); // Okay if the move constructor A(A &&other) is
                               // implemented, but that's somehow tedious.
```

So what we usually do is:
```
A *a = new A();
PtrVector<A> a_vec;
a_vec.push_back(a); // Take ownership of a.
```